File: reconhecedorLLC.py



def getSimbolo(simbolo, regras):
	#print('getSimbolo')
	lista_simbolos = []
	for regra in regras:
		if(regra[1] == simbolo):
			lista_simbolos.append(regra[0])
			#print('Encontrado: ' + regra[0] + '->' + simbolo)
			
	return lista_simbolos

def remove_repetidos(lista):
    l = []
    for i in lista:
        if i not in l:
            l.append(i)
    return l


def main():
	iterador = 1
	while 1:
		regras = []
		palavras = []
		try:
			simbolo_inicial = raw_input()

			conjunto_nao_terminais = raw_input()
			conjunto_terminais = raw_input()

			cond_parada = 1
			while(cond_parada == 1):
				regra = raw_input()
				if (regra == '# -> #'):
					cond_parada = 0
				else:
					regras.append([regra[0], regra.split("-> ")[1]])


			cond_parada = 1
			while(cond_parada == 1):
				palavra = raw_input()
				if (palavra == '#'):
					cond_parada = 0
				else:
					palavras.append(palavra)

			#print("Simbolo Inicial: ")
			#print(simbolo_inicial)
			#print("Nao terminais: ")
			#print(conjunto_nao_terminais)
			#print("Terminais: ")
			#print(conjunto_terminais)
			#print("Regras: ")
			#print(regras)
			#print("Palavras: ")
			#print(palavras)


			#Percorrer o vetor de palavras para verificar cada uma delas
			print('\nInstancia ' + str (iterador) )
			for palavra in palavras:
				triangulo = []
				triangulo.append('X')
				triangulo.append([])

				triangulo[1].append('X')
				for i in range(1, len(palavra)+1):
					lista = getSimbolo(palavra[i-1],regras)
					#print('Lista obtida para o terminal ' + palavra[i] + ': ')
					#print(lista)
					if not lista:
							lista.append('*')
					triangulo[1].append(lista)

				#print('Triangulo obtido com a subtituicao dos terminais: ')
				#print(triangulo)

				tamanho_palavra= len(palavra)
				#print(' s = len(triangulo[0]) = ' + str(s) )
				for linha in range(2,tamanho_palavra+1):
					#print("l = " + str(linha))
					triangulo.append([])
					triangulo[linha].append('X')
					#print(len(palavra))
					for coluna in range(1, ((tamanho_palavra-linha)+2)):
						#print("k = " + str(coluna))
	
						lista_simbolos = []

						for aux in range(1, linha):
							#print('L= ' + str(linha) + 'K= ' + str(coluna) + 'F= ' + str(aux))

							for item_a in triangulo[linha-aux][coluna]:
								
								for item_b in triangulo[aux][(linha-aux)+coluna]:
									#print('item_a: ' + str(item_a) + ' item_b: ' + str(item_b))


									simbolo = item_a + item_b

									#print('Simbolo: ')
									#print(simbolo)


									lista = getSimbolo(simbolo, regras)
									lista_simbolos = lista_simbolos + lista
									#print('Lista obtida para ' + simbolo + ': ')
									#print(lista_simbolos)
									

									#print('------------------')
									#print( str(triangulo[0][0]))
									#print( str(simbolo_inicial))
									#print(str(triangulo[0][0]) == str(simbolo_inicial))
									#print('------------------')

						if not lista_simbolos:
							lista_simbolos.append('*')

						lista_simbolos = remove_repetidos(lista_simbolos)
						#print('Lista adicionada ao triangulo')
						#print(lista_simbolos)
						triangulo[linha].append(lista_simbolos)

					#print('Triangulo apos a iteracao ' + str(i) + ': ')
					#rseqn = triangulo[:]
					#rseqn.reverse()

					#for item in rseqn:
						#print(item)

				#print('Triangulo gerado, caso o simpobolo inicial esteja na posicao 0 do triangulo a palavra pertence a linguagem: ')
				#print(triangulo[tamanho_palavra])

				pertence = False
				for item in triangulo[tamanho_palavra][1]:
					if(item == simbolo_inicial):
						pertence = True

				if (pertence == True):
					print( palavra + ' e uma palavra valida')
				else:
					print( palavra + ' nao e uma palavra valida')
		except EOFError:
			break
		iterador+=1

if __name__ == "__main__":
	main()

File: reconhecedorLLC.cpp
#include <iostream>
#include <string>
#include <vector>
#include <stdio.h>
#include <string.h>

using std::string;
using std::getline;
using namespace std;

string remove_repetidos(string lista_simbolos){
    string lista_sem_repetidos = "";

    for (int i =0; i < lista_simbolos.size(); i++){
        if (lista_sem_repetidos.find(lista_simbolos[i]) == -1){
            lista_sem_repetidos += lista_simbolos[i];
        }
    }
    return lista_sem_repetidos;
}

int main() {

    int iterador= 1;
    char regraE[10];
    char regraD[10];
    char palavra[51];
    char simbolo_inicial[5];
    char conjunto_nao_terminais[100], conjunto_terminais[100];
    string triangulo[60][60], simbolo;
    int iterador_terminais, iterador_palavras;

    int cond_parada;

    scanf("%s", simbolo_inicial);
    while(1){
        //ler simboloinicial, conjuntos de terminais e conjunto de não terminais em forma de string
        scanf("%s", conjunto_terminais);
        scanf("%s", conjunto_nao_terminais);
        vector<string> regras_E;
        vector<string> regras_D;
        vector<string> palavras;
        //ler regras enquanto nao encontrar uma regra do tipo '# -> #'
        cond_parada = 1;
        while(cond_parada == 1){
            scanf("%s -> %s", regraE, regraD );
            if( regraE[0] == '#'){ 
                cond_parada = 0;
            } else {
                regras_E.push_back(regraE);
                regras_D.push_back(regraD);
            }
        }
        //ler palavras enquanto nao encontrar uma palavra do tipo '#'
        cond_parada = 1;
        while(cond_parada == 1){
            scanf("%s", palavra);
            if (palavra[0] == '#'){
                cond_parada = 0;
            } else{
                palavras.push_back(palavra);
            }
        }
    
        cout << "Instancia " << iterador << endl;
        //laco de repeticao para percorrer todas as palavras
        for (iterador_palavras=0; iterador_palavras < palavras.size(); iterador_palavras++) {

            //laco de repeticao para percorrer todos os caracteres de uma determinada palavra
            for (iterador_terminais =1; iterador_terminais <= palavras[iterador_palavras].size(); iterador_terminais++){

                //determinado caracter da palavra    
                char simbolo= palavras[iterador_palavras][iterador_terminais-1];

                //pesquisar NAO TERMINAIS que geram o simbolo
                string lista_simbolos = "";
                //laco de repeticao para percorrer todas as regras
                for (int iterador_regras=0; iterador_regras < regras_E.size(); iterador_regras++) {
                    //se um determinado simbolo a direita da regra for igual ao simbolo terminal
                    if(regras_D[iterador_regras][0] == simbolo){
                        //adiciona na lista de simbolos o simbolo a esquerda daquela regra
                        lista_simbolos += regras_E[iterador_regras][0];
                    }
                }
                if(lista_simbolos.size() == 0) {
                    lista_simbolos = "*";
                }
                //Adiciona na primeira linha do triangulo a lista de simbolos encontrados que geram o determinado terminal  
                triangulo[1][iterador_terminais] = lista_simbolos;
            }

            int tamanho_palavra= palavras[iterador_palavras].size();
            string lista_simbolos;

            //laco para percorrer as linha do triangulo
            for(int linha= 2; linha < tamanho_palavra+1; linha++) {
                //laco para percorrer as colunas do triangulo
                for(int coluna= 1; coluna < (tamanho_palavra-linha+2); coluna++){
                    lista_simbolos = "";

                    for( int aux = 1; aux < linha; aux++) {

                        //lacos de repeticao para fazer as combinacoes de simbolos em posicoes do tringulo que possuem mais de um simbolo
                        for (int item1 = 0; item1 < triangulo[linha-aux][coluna].size(); item1++) {
                            for (int item2 = 0; item2 < triangulo[aux][(linha-aux)+coluna].size(); item2++) {
                                //concatena os simbolos de duas posicoes do triangulo
                                simbolo = "";
                                simbolo += triangulo[linha-aux][coluna][item1];
                                simbolo += triangulo[aux][(linha-aux)+coluna][item2];
                                //laco de repeticao para percorrer todas as regras
                                for (int iterador_regras=0; iterador_regras < regras_E.size(); iterador_regras++) {
                                    //se um determinado simbolo a direita da regra for igual ao simbolo gerado 
                                    if(regras_D[iterador_regras] == simbolo){
                                        //adiciona na lista de simbolos o simbolo a esquerda daquela regra
                                        lista_simbolos += regras_E[iterador_regras][0];
                                    }
                                }
                            }
                        }
                    }

                    if(lista_simbolos.size() == 0) {
                        lista_simbolos = "*";
                    } else {
                        //retira as variaveis repetidas da lista de simbolos
                        lista_simbolos = remove_repetidos(lista_simbolos);   
                    }
                    //Adiciona na determinada linha do triangulo a lista de simbolos encontrados
                    triangulo[linha][coluna] = lista_simbolos;
                }
            }
            int pertence = 0;
            //percorre a lista de simbolos da posicao do topo do trinagulo
            for (int item = 0; item < triangulo[tamanho_palavra][1].size(); item++) {
                //se algum dos simbolos da lista de simbolos da posicao do topo do trinagulo for igual ao simbolo inicial a palavra e valida
                if(triangulo[tamanho_palavra][1][item] == simbolo_inicial[0]){
                    pertence = 1;
                }
            }
            if (pertence == 1){
                cout <<  palavras[iterador_palavras] <<  " e uma palavra valida" << endl;
            }else{
                cout << palavras[iterador_palavras] << " nao e uma palavra valida" << endl;
            }
        }
        iterador++;
        if(scanf("%s", simbolo_inicial) != EOF){
            cout << endl;
        } else {
            break;
        }
    }
    return 0;
}
